### ISSUE TITLE
Multithreaded download for rclone

### ISSUE THREAD

**Issue by ncw** [ncw/rclone#2252]
Rclone has all the parts for doing multithreaded downloads - every remote can read a chunk out of the middle of a file.

It would be a great idea if `operations.Copy` could optionally do multipart downloads.

In order to fit in the streaming architecture of rclone the destination remote would have to support uploading a file in parts. 

This could possibly co-ordinate with the multipart upload of existing remotes
  * make a Multipart upload interface
  * this could return an object to control the upload, give info about the chunks, etc
  * it would need to be able to support uploads in parallel

This might be an opportunity for the multipart uploads to be co-ordinated at a higher level also.

This could be supported by the local remote and any which supported out of order multipart uploads.

Any upload which was going to be multipart could use multipart download (if capable of out of order?  or maybe have an in-order mode).

This would also help with #87

**Comment by daniel-loader** [comment]
Somewhat similar to how there's workers grabbing chunks for the cache simultaneously?

**Comment by anirupdutta** [comment]
@ncw We recently came up with an issue where we tried downloading a large file(700 GB+) from S3 locally and it would time out. Supporting multi thread downloads would be a great feature add. Is it something that is planned soon?

**Comment by anirupdutta** [comment]
@ncw Just to follow up, I see it is planned for v1.48? Do we think we will get it done by v1.48? Also if I want to contribute, do you think we can start with support for S3 and then expand to other services or do you think we need to have a generic solution which supports all the providers?

**Comment by ncw** [comment]
OK I had a good think about this and decided that we can only sensibly make this work for the local backend for the moment which makes the implementation easier.

Here is a test to try:

https://beta.rclone.org/branch/v1.47.0-024-g770c9624-fix-2252-multipart-download-beta/ (uploaded in 15-30 mins)

This has two flags to control it

      --multi-thread-cutoff SizeSuffix   Use multi-thread downloads for files above this size. (default 100M)
      --multi-thread-streams int         Max number of streams to use for multi-thread downloads. (default 4)

rclone will use more threads the more multiples of the --multi-thread-cutoff the file size is, up to --multi-thread-streams

So questions I'd like some input on
- does it work?
- is the interface OK (the two command line variables)
- what should the defaults be?

if you use `-vv` you'll see debugging for the multithreaded downloads which will look like this

```
rclone copyto 10M 10M.copy -vv --bwlimit 1M -P --multi-thread-cutoff 1M --multi-thread-streams 4
2019/04/24 20:46:24 DEBUG : rclone: Version "v1.47.0-DEV" starting with parameters ["rclone" "copyto" "10M" "10M.copy" "-vv" "--bwlimit" "1M" "-P" "--multi-thread-cutoff" "1M" "--multi-thread-streams" "4"]
2019/04/24 20:46:24 DEBUG : Using config file from "/home/ncw/.rclone.conf"
2019/04/24 20:46:24 INFO  : Starting bandwidth limiter at 1MBytes/s
2019-04-24 20:46:24 DEBUG : 10M: Couldn't find file - need to transfer
2019-04-24 20:46:24 DEBUG : 10M: Starting multi-thread copy with 4 parts of size 2.500M
2019-04-24 20:46:24 DEBUG : 10M: multi-thread copy: stream 4/4 (8060928-10485766) size 2.313M starting
2019-04-24 20:46:24 DEBUG : 10M: multi-thread copy: stream 2/4 (2686976-5373952) size 2.562M starting
2019-04-24 20:46:24 DEBUG : 10M: multi-thread copy: stream 1/4 (0-2686976) size 2.562M starting
2019-04-24 20:46:24 DEBUG : 10M: multi-thread copy: stream 3/4 (5373952-8060928) size 2.562M starting
2019-04-24 20:46:34 DEBUG : 10M: multi-thread copy: stream 4/4 (8060928-10485766) size 2.313M finished
2019-04-24 20:46:34 DEBUG : 10M: multi-thread copy: stream 3/4 (5373952-8060928) size 2.562M finished
2019-04-24 20:46:34 DEBUG : 10M: multi-thread copy: stream 2/4 (2686976-5373952) size 2.562M finished
2019-04-24 20:46:34 DEBUG : 10M: multi-thread copy: stream 1/4 (0-2686976) size 2.562M finished
2019-04-24 20:46:34 DEBUG : 10M: Finished multi-thread copy with 4 parts of size 2.500M
2019-04-24 20:46:35 INFO  : 10M: Multi-thread Copied (new)
```

**Comment by ErAzOr2k** [comment]
does this also work with mount?

**Comment by DurvalMenezes** [comment]
Hello @ncw, 
> Here is a test to try:
> https://beta.rclone.org/branch/v1.47.0-024-g770c9624-fix-2252-multipart-download-beta/ (uploaded in 15-30 mins)

The above link is returning error 404... so either your 15-30 mins got stretched quite a bit, or something is wrong ;-)

Cheers, 
-- Durval.

**Comment by ncw** [comment]
Ah, that one failed the tests...

Here is a new one

https://beta.rclone.org/branch/v1.47.0-028-g1de11062-fix-2252-multipart-download-beta/ (uploaded in 15-30 mins)

Cross fingers!

**Comment by DurvalMenezes** [comment]
Just tested it, and it's working great here! full(er) report at https://forum.rclone.org/t/multi-threaded-downloads-comments-and-testers-needed/9721/24?u=durval

**Comment by DurvalMenezes** [comment]
@ncw, one question: I tested it with a single-file download from an SFTP remote, but it seems to have simply ignored `--multi-thread-cutoff=1 --multi-thread-streams=4` and proceeded to copy the file in the "normal" single-threaded way. Doesn't the SFTP remote support multithreading operation? By looking at the SFTP "API" (as exposed by golang's  `sftp` package), it seems to support reading from/writing to arbitrary offsets:  https://godoc.org/github.com/pkg/sftp#File.Seek 
Is anything else required for multithreaded operation?

**Comment by AndShy** [comment]
@DurvalMenezes 
are you sure that the problem isn't in --multi-thread-cutoff=1 without suffix ? as i understand without suffix default kBytes

@ncw 
i tested new function with OneDrive, Team GDrive and mega.nz. looks like it's working pretty good, but mega.nz for stable transfer with full bandwith require around 20 streams. otherwise speed jumping from 40-60  to 100 Mb (channel - 100 Mb, 1 stream download - 8.7 Mb)
Really looking forward to multistream uploading implementation, cause usually uploading speed much lower than downloading. 
Thank you for great software!

**Comment by DurvalMenezes** [comment]
@AndShy, @ncw:
> I'm pretty sure it was not. But I will pay attention to that on my next tests, in order to discard this hypothesis completely.

As expected, the problem does not seem to be connected with using a prefix; please see [my post in the forum](https://forum.rclone.org/t/multi-threaded-downloads-comments-and-testers-needed/9721/30?u=durval).

**Comment by ncw** [comment]
I wrote in the forum what I think is the explanation:

Multi thread copy will only work to the local backend for the time being. It could in theory work for some other backends but it is really complicated and I haven’t got my head round how to do it!

**Comment by AndShy** [comment]
@ncw 
better concentrate on multi-thread uploading. it's more important :)
file transfering between remote backends primarily must be their own problem or must be carried out by third side server / service (like multcloud.com)

**Comment by ncw** [comment]
As I said above I don't think google drive can do multi-thread uploading.

However to double check my understanding I asked a question on stack overflow: https://stackoverflow.com/q/56076208/164234

Mega definitely could, but I haven't implemented it. I'm not sure about onedrive.

**Comment by AndShy** [comment]
looks like need do this "manualy" by somehow split large file on local side and check for splited files on remote side.

**Comment by DurvalMenezes** [comment]
Howdy @ncw, @AndShy:

> Multi thread copy will only work to the local backend for the time being. It could in theory work for some other backends but it is really complicated and I haven’t got my head round how to do it!

I may have found a way around that, via a separate `rclone mount` ([see here for complete details](https://forum.rclone.org/t/multi-threaded-downloads-comments-and-testers-needed/9721/34?u=durval))

> As I said above I don't think google drive can do multi-thread uploading.

> looks like need do this "manualy" by somehow split large file on local side and check for splited files on remote side.

The same trick *may* perhaps (more dubiously) be used to work around that; again, see my post in the forum linked above.

Cheers, 
-- Durval.

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.48.

Thank you all for testing.
